### ISSUE TITLE
error in backward function in log softmax

### ISSUE THREAD

**Issue by mrityunjay-tripathi** [mlpack/mlpack#2360]
#### Issue description
The backward function of ```log_softmax.hpp``` doesn't give the correct output.

#### Steps to reproduce
In mlpack:
```
#include <bits/stdc++.h>
#include <mlpack/core.hpp>
#include <mlpack/methods/ann/layer/log_softmax.hpp>

using namespace std;
using namespace mlpack::ann;
using namespace arma;

int main()
{
  arma::mat input = arma::mat("0.1; 0.9");
  arma::mat output, gy = arma::ones(2, 1), g;
  // May be I am not using 'gy' in correct way.

  LogSoftMax<> module;
  module.Forward(input, output);
  output.print("Forward:");
  module.Backward(output, gy, g);
  g.print("Backward:");
}
```
Output:
```
Forward:
  -1.1711
  -0.3711
Backward:
  1.31003 
  1.68997
```

#### Expected behavior

In PyTorch
```
import torch
import torch.nn as nn
input = torch.tensor([0.1, 0.9], requires_grad = True)
output = nn.LogSoftmax()(input)
print("Forward:", output)
output.backward(torch.ones(2))
print("Backward:", input.grad)
```
Output:
```
Forward: tensor([-1.1711, -0.3711], grad_fn=<LogSoftmaxBackward>)
Backward: tensor([ 0.3799, -0.3799])
```

Probably, the code in the backward function of ```log_softmax.hpp``` adds ```gy``` in place of the [Kronecker-Delta](https://en.wikipedia.org/wiki/Kronecker_delta) function.
Or am I doing something fundamentally wrong?

**Comment by mrityunjay-tripathi** [comment]
Alos, log_softmax doesn't pass ```JacobianTest``` when I tested it locally as the error was around 0.8.

**Comment by nishantkr18** [comment]
Hey everyone! is this issue resolved?

**Comment by mrityunjay-tripathi** [comment]
Hello @nishantkr18, yes the confusion was sorted out. Looks like previous comments are removed?!? [This discussion](https://discuss.pytorch.org/t/does-nllloss-handle-log-softmax-and-softmax-in-the-same-way/8835) resolved "my" confusion.
